aws/resource_aws_config_remediation_configuration.go

```diff
@@ -0,0 +1,217 @@
+	remediationConfigurationInput := configservice.RemediationConfiguration{
+		ConfigRuleName: aws.String(name),
+	}
+
+	if v, ok := d.GetOk("parameters"); ok {
+		remediationConfigurationInput.Parameters = expandConfigRemediationConfigurationParameters(v.(*schema.Set))
+	}
+
+	if v, ok := d.GetOk("resource_type"); ok && v.(string) != "" {
+		remediationConfigurationInput.ResourceType = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("target_id"); ok && v.(string) != "" {
+		remediationConfigurationInput.TargetId = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("target_type"); ok && v.(string) != "" {
+		remediationConfigurationInput.TargetType = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("target_version"); ok && v.(string) != "" {
+		remediationConfigurationInput.TargetVersion = aws.String(v.(string))
+	}
+
+	remediationConfigurations := make([]*configservice.RemediationConfiguration, 1)
+	remediationConfigurations = append(remediationConfigurations, &remediationConfigurationInput)
+
+	input := configservice.PutRemediationConfigurationsInput{
+		RemediationConfigurations: remediationConfigurations,
+	}
+	log.Printf("[DEBUG] Creating AWSConfig remediation configuration: %s", input)
+	err := resource.Retry(2*time.Minute, func() *resource.RetryError {
+		_, err := conn.PutRemediationConfigurations(&input)
+		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok {
+				if awsErr.Code() == "InsufficientPermissionsException" {
+					// IAM is eventually consistent
+					return resource.RetryableError(err)
+				}
+			}
+
+			return resource.NonRetryableError(fmt.Errorf("Failed to create AWSConfig remediation configuration: %s", err))
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	d.SetId(name)
+
+	log.Printf("[DEBUG] AWSConfig config remediation configuration for rule %q created", name)
+
+	return resourceAwsConfigRemediationConfigurationRead(d, meta)
+}
+
+func resourceAwsConfigRemediationConfigurationRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).configconn
+	out, err := conn.DescribeRemediationConfigurations(&configservice.DescribeRemediationConfigurationsInput{
+		ConfigRuleNames: []*string{aws.String(d.Id())},
+	})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "NoSuchConfigRuleException" {
+			log.Printf("[WARN] Config Rule %q is gone (NoSuchConfigRuleException)", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	numberOfRemediationConfigurations := len(out.RemediationConfigurations)
+	if numberOfRemediationConfigurations < 1 {
+		log.Printf("[WARN] No Remediation Configuration for Config Rule %q (no remediation configuration found)", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	log.Printf("[DEBUG] AWS Config remediation configurations received: %s", out)
+
+	remediationConfigurations := out.RemediationConfigurations
+	d.Set("remediationConfigurations", flattenRemediationConfigurations(remediationConfigurations))
```

Comment:
I'll have to put this error handling code in the `flattenRemediationConfigurations`, then, as this code handles lists of `RemediationConfiguration` objects.
Unless you mean I have to handle `RemediationConfiguration` one by one?

Reply:
The more I read this specific review, my code, and AWS documentation (website and `man`), the more I feel like there is something to clarify before getting forward :
`Remediation Configuration` objects can only be identified by their corresponding `Config Rule`, which makes sense as they dannot exist without a `Config Rule` on top of them.
`aws` CLI comand reflects that fact :
```
$ aws configservice help
...
AVAILABLE COMMANDS
...
delete-remediation-configuration
...
describe-remediation-configurations
...
put-remediation-configurations
...
```

```
$ aws configservice describe-config-rules help
...
SYNOPSIS
            describe-config-rules
          [--config-rule-names <value>]
          [--cli-input-json <value>]
          [--starting-token <value>]
          [--max-items <value>]
          [--generate-cli-skeleton <value>]

OPTIONS
...
```
Which appears the same way in the GO API this provider is using to communicate with AWS, if I understood correctly.

At this point I'm not sure I wrote a code that matches these behaviors : maybe it would be best to only handle lists of Remediation Configuraion objects, be it when reading or writing them, instead of allowing to write a single `Remediation Configuration` object outside of any kind of collection.